File: my_openclaw/mini_agent_ts/src/agent.ts
/**
 * agent.ts — Mini Agent main entry point
 *
 * Borrowed from openclaw's agent execution pipeline:
 *   - src/agents/agent-command.ts           (agentCommand — the main orchestrator)
 *   - src/agents/command/attempt-execution.ts (runAgentAttempt — LLM call)
 *   - src/agents/model-selection.ts          (model resolution)
 *   - src/agents/defaults.ts                 (DEFAULT_PROVIDER, DEFAULT_MODEL)
 *
 * openclaw's full pipeline (simplified):
 *   ┌─────────────────────────────────────────────────────────────────┐
 *   │ 1. agentCommand()                                              │
 *   │    ├─ loadConfig() → resolve secrets → setRuntimeConfigSnapshot│
 *   │    ├─ resolveSession() → find/create session                   │
 *   │    ├─ resolveConfiguredModelRef() → pick model + provider      │
 *   │    ├─ buildAllowedModelSet() → validate model allowlist        │
 *   │    ├─ resolveSessionTranscriptFile() → session file path       │
 *   │    └─ runWithModelFallback()                                   │
 *   │         └─ runAgentAttempt()                                   │
 *   │              ├─ isCliProvider? → runCliAgent()                  │
 *   │              └─ else → runEmbeddedPiAgent()                    │
 *   │                   └─ @mariozechner/pi-ai (Anthropic/OpenAI SDK)│
 *   │                        └─ fetch POST /chat/completions         │
 *   │                             └─ SSE stream → token callbacks    │
 *   │ 2. deliverAgentCommandResult()                                 │
 *   │    └─ send reply via channel (Slack/Discord/Telegram/etc.)     │
 *   └─────────────────────────────────────────────────────────────────┘
 *
 * Our mini version:
 *   ┌─────────────────────────────────────────────────────────────────┐
 *   │ 1. loadConfig(HAI_WOA.json)                                    │
 *   │ 2. resolveModel(alias) → API model name                       │
 *   │ 3. Read user input from stdin                                  │
 *   │ 4. MessageHistory.addUser(input)                               │
 *   │ 5. streamChat() → POST /chat/completions (SSE)                │
 *   │ 6. Token callback → process.stdout.write()                    │
 *   │ 7. MessageHistory.addAssistant(fullResponse)                   │
 *   │ 8. Loop back to step 3                                        │
 *   └─────────────────────────────────────────────────────────────────┘
 */

import { createInterface } from "node:readline";
import { loadConfig, resolveModel, dumpConfig, type AgentConfig } from "./config.js";
import { MessageHistory } from "./message.js";
import { streamChat } from "./http-client.js";

// ── CLI argument parsing ──────────────────────────────────────────
// Minimal arg parsing — mirrors openclaw's commander-based CLI
// but stripped to essentials.

interface CliArgs {
  configPath: string;
  model?: string;
}

function parseArgs(): CliArgs {
  const args = process.argv.slice(2);
  let configPath = "../HAI_WOA.json";
  let model: string | undefined;

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case "-c":
      case "--config":
        configPath = args[++i] ?? configPath;
        break;
      case "-m":
      case "--model":
        model = args[++i];
        break;
      case "-h":
      case "--help":
        console.log(`
Usage: mini-agent-ts [options]

Options:
  -c, --config <path>   Path to HAI_WOA.json config file (default: ../HAI_WOA.json)
  -m, --model <name>    Model name or alias (default: from config)
  -h, --help            Show this help message

Commands (in chat):
  /model <name>         Switch to a different model
  /clear                Clear conversation history
  /history              Show conversation history
  quit                  Exit the agent
`);
        process.exit(0);
    }
  }

  return { configPath, model };
}

// ── REPL loop ─────────────────────────────────────────────────────
// Interactive read-eval-print loop. Mirrors openclaw's agent command
// flow but without channels — reads directly from stdin.

async function runAgent(): Promise<void> {
  // ── Step 1: Load configuration ──────────────────────────────
  // Mirrors: loadConfig() → resolveCommandSecretRefsViaGateway()
  const cliArgs = parseArgs();
  let cfg: AgentConfig;
  try {
    cfg = loadConfig(cliArgs.configPath);
  } catch (err) {
    console.error(`[agent] Failed to load config: ${err}`);
    process.exit(1);
  }
  dumpConfig(cfg);

  // ── Step 2: Resolve model ───────────────────────────────────
  // Mirrors: resolveConfiguredModelRef() → normalizeModelRef()
  let currentModel = resolveModel(cfg, cliArgs.model);

  // ── Step 3: Initialize message history ──────────────────────
  // Mirrors: SessionManager.open(sessionFile)
  const systemPrompt =
    "You are a helpful assistant. Be concise and clear in your responses.";
  let history = new MessageHistory(systemPrompt);

  // ── Banner ──────────────────────────────────────────────────
  console.log("");
  console.log("  ╔═══════════════════════════════════════════╗");
  console.log("  ║   🤖 Mini Agent TS (openclaw-inspired)    ║");
  console.log("  ╠═══════════════════════════════════════════╣");
  console.log(`  ║  Model: ${currentModel.padEnd(32)} ║`);
  console.log("  ║  Type 'quit' or Ctrl-D to exit            ║");
  console.log("  ║  Type '/model <name>' to switch model     ║");
  console.log("  ║  Type '/clear' to reset conversation      ║");
  console.log("  ║  Type '/history' to show message history  ║");
  console.log("  ╚═══════════════════════════════════════════╝");
  console.log("");

  // ── Step 4: REPL loop ───────────────────────────────────────
  // Mirrors: the agent command's message → runAgentAttempt() → deliver cycle
  const rl = createInterface({
    input: process.stdin,
    output: process.stdout,
    prompt: "You > ",
  });

  // Line queue for proper async handling.
  // In pipe mode, readline fires all "line" events synchronously before
  // any async handler can process them. We buffer lines in a queue and
  // resolve them one at a time via askLine().
  const lineQueue: string[] = [];
  let lineResolve: ((line: string | null) => void) | null = null;
  let eofReached = false;

  rl.on("line", (line) => {
    if (lineResolve) {
      const resolve = lineResolve;
      lineResolve = null;
      resolve(line);
    } else {
      lineQueue.push(line);
    }
  });

  rl.on("close", () => {
    eofReached = true;
    if (lineResolve) {
      const resolve = lineResolve;
      lineResolve = null;
      resolve(null);
    }
  });

  const askLine = (): Promise<string | null> => {
    // Drain buffered lines first
    if (lineQueue.length > 0) {
      return Promise.resolve(lineQueue.shift()!);
    }
    if (eofReached) {
      return Promise.resolve(null);
    }
    return new Promise((resolve) => {
      lineResolve = resolve;
      rl.prompt();
    });
  };

  let running = true;
  while (running) {
    const line = await askLine();
    if (line === null) {
      // EOF (Ctrl-D or pipe exhausted)
      console.log("[Goodbye!]");
      break;
    }

    const input = line.trim();

    // ── Handle empty input ──────────────────────────────────
    if (!input) {
      continue;
    }

    // ── Handle quit ─────────────────────────────────────────
    if (input.toLowerCase() === "quit" || input.toLowerCase() === "exit") {
      console.log("[Goodbye!]");
      running = false;
      continue;
    }

    // ── Handle /model command ───────────────────────────────
    // Mirrors: openclaw's model override via session store
    if (input.startsWith("/model ")) {
      const modelArg = input.slice(7).trim();
      if (!modelArg) {
        console.log(`[Current model: ${currentModel}]`);
      } else {
        currentModel = resolveModel(cfg, modelArg);
      console.log(`[Switched to model: ${currentModel}]`);
      }
      console.log("");
      continue;
    }

    // ── Handle /clear command ───────────────────────────────
    if (input === "/clear") {
      history = new MessageHistory(systemPrompt);
      console.log("[Conversation cleared]");
      console.log("");
      continue;
    }

    // ── Handle /history command ─────────────────────────────
    if (input === "/history") {
      const msgs = history.getMessages();
      console.log(`\n[Message history: ${msgs.length} messages]`);
      for (const msg of msgs) {
        const preview =
          msg.content.length > 80
            ? msg.content.slice(0, 80) + "..."
            : msg.content;
        console.log(`  [${msg.role}] ${preview}`);
      }
      console.log("");
      continue;
    }

    // ── Step 5: Add user message to history    // Mirrors: prependInternalEventContext(message) → SessionManager.appendMessage()
    history.addUser(input);

    // ── Step 6: Call LLM API ────────────────────────────────
    // Mirrors: runAgentAttempt() → runEmbeddedPiAgent() → pi-ai fetch
    process.stdout.write("Assistant > ");

    let fullResponse = "";
    try {
      await streamChat({
        baseUrl: cfg.baseUrl,
        apiKey: cfg.apiKey,
        model: currentModel,
        messages: history.getMessages(),
        onToken: (token: string, done: boolean) => {
          if (done) {
            // Stream finished — add newline
            process.stdout.write("\n");
            return;
          }
          // ── Step 7: Stream tokens to stdout ───────────────
          // Mirrors: emitAcpAssistantDelta() → channel delivery
          fullResponse += token;
          process.stdout.write(token);
        },
      });

      // ── Step 8: Save assistant response to history ────────
      // Mirrors: SessionManager.appendMessage({ role: "assistant" })
      if (fullResponse) {
        history.addAssistant(fullResponse);
      }
    } catch (err) {
      console.error(`\n[Error: ${err instanceof Error ? err.message : String(err)}]`);
    }

    console.log("");
  }

  rl.close();
}

// ── Entry point ───────────────────────────────────────────────────

runAgent().catch((err) => {
  console.error(`[fatal] ${err}`);
  process.exit(1);
});

File: my_openclaw/mini_agent/agent.c
/*
 * agent.c - Mini Agent: the complete input-to-output pipeline
 *
 * ═══════════════════════════════════════════════════════════════
 *  Architecture Overview (mirrors openclaw's core flow):
 * ═══════════════════════════════════════════════════════════════
 *
 *  ┌─────────────┐
 *  │  HAI_WOA.json│  ← Configuration (model, API key, base URL)
 *  └──────┬──────┘
 *         │ config_load()
 *         ▼
 *  ┌─────────────┐
 *  │  agent_config│  ← Parsed config struct
 *  └──────┬──────┘
 *         │
 *         ▼
 *  ┌─────────────────────────────────────────────────────┐
 *  │                   MAIN LOOP                         │
 *  │                                                     │
 *  │  1. Read user input from stdin                      │
 *  │  2. Add to conversation (message.c)                 │
 *  │  3. Serialize messages to JSON                      │
 *  │  4. POST to /chat/completions with SSE streaming    │
 *  │  5. Parse SSE "data:" lines in real-time            │
 *  │  6. Extract choices[0].delta.content                │
 *  │  7. Print tokens as they arrive                     │
 *  │  8. Accumulate full response                        │
 *  │  9. Add assistant reply to conversation history     │
 *  │ 10. Loop back to step 1                             │
 *  └─────────────────────────────────────────────────────┘
 *
 * ═══════════════════════════════════════════════════════════════
 *  Data Flow:
 * ═══════════════════════════════════════════════════════════════
 *
 *  stdin ──► conversation_add(USER) ──► conversation_to_json()
 *                                              │
 *                                              ▼
 *                                    http_stream_chat()
 *                                              │
 *                                    ┌─────────┴─────────┐
 *                                    │  libcurl POST      │
 *                                    │  SSE write_callback│
 *                                    │  parse "data:" line│
 *                                    │  json_parse(chunk) │
 *                                    │  extract .content  │
 *                                    └─────────┬─────────┘
 *                                              │
 *                                              ▼
 *                                    stream_token_cb()
 *                                              │
 *                                    ┌─────────┴─────────┐
 *                                    │  print to stdout   │
 *                                    │  accumulate reply  │
 *                                    └─────────┬─────────┘
 *                                              │
 *                                              ▼
 *                                    conversation_add(ASSISTANT)
 *                                              │
 *                                              ▼
 *                                         next turn
 *
 * ═══════════════════════════════════════════════════════════════
 *  Module Dependency:
 * ═══════════════════════════════════════════════════════════════
 *
 *  agent.c (this file)
 *    ├── config.h      → loads HAI_WOA.json
 *    │     └── json_parser.h  → parses JSON
 *    ├── message.h     → manages conversation history
 *    └── http_client.h → streams chat completions
 *          ├── json_parser.h  → parses SSE chunks
 *          └── libcurl        → HTTP transport
 */

#include "config.h"
#include "message.h"
#include "http_client.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>

/* ── Default system prompt ─────────────────────────────────────── */

#define DEFAULT_SYSTEM_PROMPT \
    "You are a helpful AI assistant. Answer concisely and clearly."

/* ── Global state for signal handling ──────────────────────────── */

static volatile int g_running = 1;

static void sigint_handler(int sig)
{
    (void)sig;
    g_running = 0;
    printf("\n[Interrupted]\n");
}

/* ── Response accumulator (used in streaming callback) ─────────── */

typedef struct {
    char   *buf;
    size_t  len;
    size_t  cap;
} response_buf_t;

static void response_init(response_buf_t *r)
{
    r->cap = 1024;
    r->len = 0;
    r->buf = malloc(r->cap);
    r->buf[0] = '\0';
}

static void response_append(response_buf_t *r, const char *s, size_t slen)
{
    while (r->len + slen + 1 > r->cap) {
        r->cap *= 2;
        r->buf = realloc(r->buf, r->cap);
    }
    memcpy(r->buf + r->len, s, slen);
    r->len += slen;
    r->buf[r->len] = '\0';
}

static void response_free(response_buf_t *r)
{
    free(r->buf);
}

/* ── Streaming callback: print tokens + accumulate ─────────────── */

static void on_token(const char *token, size_t len, int done, void *userdata)
{
    response_buf_t *resp = (response_buf_t *)userdata;

    if (done) {
        /* Stream finished */
        printf("\n");
        fflush(stdout);
        return;
    }

    /* Print token immediately (real-time streaming effect) */
    fwrite(token, 1, len, stdout);
    fflush(stdout);

    /* Accumulate for conversation history */
    response_append(resp, token, len);
}

/* ── Read a line of input from stdin ───────────────────────────── */

static char *read_input(void)
{
    printf("\033[1;36mYou > \033[0m");
    fflush(stdout);

    size_t cap = 256, len = 0;
    char *buf = malloc(cap);

    int c;
    while ((c = fgetc(stdin)) != EOF && c != '\n') {
        if (len + 1 >= cap) {
            cap *= 2;
            buf = realloc(buf, cap);
        }
        buf[len++] = (char)c;
    }

    if (c == EOF && len == 0) {
        free(buf);
        return NULL; /* EOF */
    }

    buf[len] = '\0';
    return buf;
}

/* ── Print usage ───────────────────────────────────────────────── */

static void print_usage(const char *prog)
{
    printf("Usage: %s [options]\n", prog);
    printf("Options:\n");
    printf("  -c <path>    Path to HAI_WOA.json (default: ../HAI_WOA.json)\n");
    printf("  -m <model>   Model name or alias (default: from config)\n");
    printf("  -s <prompt>  Custom system prompt\n");
    printf("  -h           Show this help\n");
}

/* ── Print banner ──────────────────────────────────────────────── */

static void print_banner(const char *model)
{
    printf("\n");
    printf("  ╔═══════════════════════════════════════════╗\n");
    printf("  ║     🤖 Mini Agent (openclaw-inspired)     ║\n");
    printf("  ╠═══════════════════════════════════════════╣\n");
    printf("  ║  Model: %-33s║\n", model);
    printf("  ║  Type 'quit' or Ctrl-D to exit            ║\n");
    printf("  ║  Type '/model <name>' to switch model     ║\n");
    printf("  ╚═══════════════════════════════════════════╝\n");
    printf("\n");
}

/* ═══════════════════════════════════════════════════════════════ */
/*                          MAIN                                  */
/* ═══════════════════════════════════════════════════════════════ */

int main(int argc, char *argv[])
{
    /* ── Parse command-line arguments ──────────────────────────── */
    const char *config_path   = "../HAI_WOA.json";
    const char *model_arg     = NULL;
    const char *system_prompt = DEFAULT_SYSTEM_PROMPT;

    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "-c") == 0 && i + 1 < argc)
            config_path = argv[++i];
        else if (strcmp(argv[i], "-m") == 0 && i + 1 < argc)
            model_arg = argv[++i];
        else if (strcmp(argv[i], "-s") == 0 && i + 1 < argc)
            system_prompt = argv[++i];
        else if (strcmp(argv[i], "-h") == 0) {
            print_usage(argv[0]);
            return 0;
        }
    }

    /* ── Step 1: Load configuration from HAI_WOA.json ─────────── */
    agent_config_t cfg;
    if (config_load(&cfg, config_path) != 0) {
        fprintf(stderr, "Failed to load config from: %s\n", config_path);
        return 1;
    }
    config_dump(&cfg);

    /* Resolve which model to use */
    const char *model = config_resolve_model(&cfg, model_arg);

    /* ── Step 2: Initialize subsystems ─────────────────────────── */
    http_init();
    signal(SIGINT, sigint_handler);

    /* ── Step 3: Initialize conversation with system prompt ────── */
    conversation_t conv;
    conversation_init(&conv, system_prompt);

    print_banner(model);

    /* ── Step 4: Main interaction loop ─────────────────────────── */
    while (g_running) {
        /* 4a. Read user input */
        char *input = read_input();
        if (!input) {
            printf("\n[EOF - Goodbye!]\n");
            break;
        }

        /* Skip empty input */
        if (input[0] == '\0') {
            free(input);
            continue;
        }

        /* Check for quit command */
        if (strcmp(input, "quit") == 0 || strcmp(input, "exit") == 0) {
            free(input);
            printf("[Goodbye!]\n");
            break;
        }

        /* Check for /model command to switch models */
        if (strncmp(input, "/model ", 7) == 0) {
            const char *new_model = config_resolve_model(&cfg, input + 7);
            model = new_model;
            printf("[Switched to model: %s]\n\n", model);
            free(input);
            continue;
        }

        /* 4b. Add user message to conversation */
        conversation_add(&conv, ROLE_USER, input);
        free(input);

        /* 4c. Serialize conversation to JSON */
        char *messages_json = conversation_to_json(&conv);

        /* 4d. Stream the response */
        printf("\033[1;32mAssistant > \033[0m");
        fflush(stdout);

        response_buf_t resp;
        response_init(&resp);

        int rc = http_stream_chat(cfg.base_url, cfg.api_key,
                                  model, messages_json,
                                  on_token, &resp);

        free(messages_json);

        if (rc != 0) {
            fprintf(stderr, "\n[Error: request failed]\n");
            response_free(&resp);
            continue;
        }

        /* 4e. Add assistant reply to conversation history */
        if (resp.len > 0) {
            conversation_add(&conv, ROLE_ASSISTANT, resp.buf);
        }

        response_free(&resp);
        printf("\n");
    }

    /* ── Cleanup ───────────────────────────────────────────────── */
    conversation_free(&conv);
    http_cleanup();

    return 0;
}
